pageserver/src/tenant/timeline.rs

```diff
@@ -4759,31 +4705,176 @@ impl Deref for TimelineWriter<'_> {
     }
 }
 
+impl Drop for TimelineWriter<'_> {
+    fn drop(&mut self) {
+        let state = &mut *self.write_guard;
+        *state = None;
+    }
+}
+
+enum OpenLayerAction {
+    Roll,
+    Open,
+    None,
+}
+
 impl<'a> TimelineWriter<'a> {
     /// Put a new page version that can be constructed from a WAL record
     ///
     /// This will implicitly extend the relation, if the page is beyond the
     /// current end-of-file.
     pub(crate) async fn put(
-        &self,
+        &mut self,
         key: Key,
         lsn: Lsn,
         value: &Value,
         ctx: &RequestContext,
     ) -> anyhow::Result<()> {
-        self.tl.put_value(key, lsn, value, ctx).await
+        let mut buf = smallvec::SmallVec::<[u8; 256]>::new();
```

Comment:
The comment about the smallvec is still in it's original place, could move it here.

Reply:
Good spot: https://github.com/neondatabase/neon/pull/6661/commits/7a958e4ae15d6fea1c7bf7819be105040aa76e26